dart-impl/base/include/dash/dart/base/atomic.h

```diff
@@ -0,0 +1,42 @@
+/*
+ * \file atomic.h
+ *
+ * Provides atomic operations on basic integer data types.
+ * Feel free to add operations you may require.
+ *
+ * See the URL below for details on the operations:
+ * https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html#g_t_005f_005fsync-Builtins
+ *
+ * See also the file kmp_os.h of the Clang OpenMP library, which served as inspiration.
+ *
+ * TODO: Check that all compilers support these intrinsics, esp. Cray?
+ */
+
+#ifndef DASH_DART_BASE_ATOMIC_H_
+#define DASH_DART_BASE_ATOMIC_H_
+
+
+#define DART_FETCH_AND_ADD64(ptr, val)   __sync_fetch_and_add((int64_t *)(ptr), (val))
```

Comment:
I added a guard flag and an unsafe alternative. CMake errors out if thread-support is requested but the builtins are not available. 

I am aware of the EMBB code but I would like to avoid pulling in assembler code by all means because it is just not portable. Just think of moving to ARM... The `__sync` builtins are supported by all compilers we currently support. Ideally we would use the `__atomic` builtins or C11 `atomic`s (at some point) but at least the latter is still not supported by the Intel compiler at this point.